consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/IbftGossip.java

```diff
@@ -0,0 +1,166 @@
+import tech.pegasys.pantheon.consensus.ibft.ibftmessagedata.ProposalPayload;
+import tech.pegasys.pantheon.consensus.ibft.ibftmessagedata.RoundChangePayload;
+import tech.pegasys.pantheon.consensus.ibft.ibftmessagedata.SignedData;
+import tech.pegasys.pantheon.consensus.ibft.network.IbftNetworkPeers;
+import tech.pegasys.pantheon.crypto.SECP256K1.Signature;
+import tech.pegasys.pantheon.ethereum.core.Address;
+import tech.pegasys.pantheon.ethereum.p2p.api.MessageData;
+
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import com.google.common.collect.Lists;
+
+/** Class responsible for rebroadcasting IBFT messages to known validators */
+public class IbftGossip {
+  private final IbftNetworkPeers peers;
+
+  // Size of the seenMessages cache, should end up utilising 65bytes * this number + some meta data
+  private final int seenMessagesSize;
+
+  // Set that starts evicting members when it hits capacity
+  private final Set<Signature> seenMessages =
+      Collections.newSetFromMap(
+          new LinkedHashMap<Signature, Boolean>() {
+            @Override
+            protected boolean removeEldestEntry(final Map.Entry<Signature, Boolean> eldest) {
+              return size() > seenMessagesSize;
+            }
+          });
+
+  IbftGossip(final IbftNetworkPeers peers, final int seenMessagesSize) {
+    this.seenMessagesSize = seenMessagesSize;
+    this.peers = peers;
+  }
+
+  /**
+   * Constructor that attaches gossip logic to a set of peers
+   *
+   * @param peers The always up to date set of connected peers that understand IBFT
+   */
+  public IbftGossip(final IbftNetworkPeers peers) {
+    this(peers, 10_000);
+  }
+
+  /**
+   * Retransmit a given proposal payload to all IBFT nodes
+   *
+   * @param payload The signed payload to retransmit
+   * @param excludeAddresses A list of addresses to not communicate this to. Does not need to
+   *     include the payload signer, that is handled implicitly
+   * @return Boolean of whether the message was rebroadcast or ignored as it's already been
+   *     broadcast in recent memory
+   */
+  public boolean gossipProposalPayload(
+      final SignedData<ProposalPayload> payload, final Address... excludeAddresses) {
+    final ProposalMessageData messageData = ProposalMessageData.create(payload);
+    return gossipMessage(
+        payload.getSignature(), messageData, payload.getSender(), excludeAddresses);
+  }
+
+  /**
+   * Retransmit a given prepare payload to all IBFT nodes
+   *
+   * @param payload The signed payload to retransmit
+   * @param excludeAddresses A list of addresses to not communicate this to. Does not need to
+   *     include the payload signer, that is handled implicitly
+   * @return Boolean of whether the message was rebroadcast or ignored as it's already been
+   *     broadcast in recent memory
+   */
+  public boolean gossipPreparePayload(
```

Comment:
The sender is available from the message object, the signer is available from the SignedData object, the signature I need to cache is in the SignedData object, dispatching a message requires it to be a MessageData. So at the end of everything all 3 (Message, MessageData, SignedData) need to be utilised at one level or another.

I've tried writing it a few ways and I've gotten to this version because it means the IbftController only needs to worry about the transport layer addresses when excluding things and the gossiper worries about the message contents. The other versions I've written didn't end up with a clean division of responsibilities.

Reply:
I know its not ideal - but in IbftController, can we not write:
        gossiper.gossipMessage(payload.getSignature(),
            messageData,
            payload.getSender(),
            payload.getSender(), message.getConnection().getPeer().getAddress());

Inside the handleMessage()? Or something similar in the new processMessage function?

Reply:
I don't think that's an improvement. This moves the concern for identifying a message from the gossiper to the controller. Ideally I'd like to pass a `Message` into the gossiper but with the type information that's lost about it's contents and the requirements to re decode the message I compromised and picked the most descriptive object that contained as much of the information I required as it could.

Reply:
Even better:
`
  public boolean gossipMessage(
      final SignedData<?> signedMessage,
      final Message message)  {
    final Signature msgUniqueId = signedMessage.getSignature();
    if (seenMessages.contains(msgUniqueId)) {
      return false;
    } else {
      final Address receivedFrom = message.getConnection().getPeer().getAddress();
      final List<Address> excludeAddressesList =
          Lists.newArrayList(receivedFrom, signedMessage.getSender());
      peers.multicastToValidatorsExcept(message.getData(), excludeAddressesList);
      seenMessages.add(msgUniqueId);
      return true;
    }
  }
`

Reply:
This doesn't make encapsulation any better, the invoker needs to co-ordinate that the message contains that SignedData.